pcs/snmp/pcs_snmp_agent.8
.TH PCS_SNMP_AGENT "8" "April 2021" "pcs-snmp 0.9.170" "System Administration Utilities"
.SH NAME
pcs_snmp_agent \- a SNMP agent providing data about a corosync/pacemaker cluster

.SH DESCRIPTION
pcs_snmp_agent is a SNMP agent which connects to the master agent (snmpd) via agentx protocol and provides information about a corosync/pacemaker cluster.

.SH SETUP
pcs_snmp_agent doesn't work as a standalone agent. It only provides data to a master agent. To configure a snmpd daemon (from package net\-snmp) as the master agent, add line 'master agentx' into the snmpd configuration file (see snmp.conf(5)). Then master agent has to be configured to accepts requests for pcs defined MIBs. This can be done by adding line 'view systemview included .1.3.6.1.4.1.32723.100' into the snmpd config file.

.SH MIB DATA VERSIONS
.TP
.B V1 \- REDHAT\-CLUSTER\-PCS\-V1\-MIB
Provides basic information about cluster such as cluster name, list of cluster nodes and list of primitive resources.

.SH ENVIRONMENT
.TP
.B PCS_SNMP_AGENT_DEBUG=<boolean>
Set to \fBtrue\fR for advanced debugging information.
.TP
.B PCS_SNMP_AGENT_UPDATE_INTERVAL=<integer>
Time interval in seconds after which agent will update provided data.

.SH SEE ALSO
.BR pcs (8)
.BR pcsd (8)

.BR snmpd (8)
.BR snmp.conf (5)


CHANGELOG.md
# Change Log

## [Unreleased]

### Added
- Elliptic curve TLS certificates are now supported in pcsd ([ghissue#123])
- Support for corosync option `totem.block_unlisted_ips` ([rhbz#1720221])

### Fixed
- Node attribute expressions are now correctly reported as not allowed in
  resource defaults rules ([rhbz#1896458])
- Upgreded to jquery 3.6.0 ([rhbz#1882291, rhbz#1886342])

### Changed
- python3-openssl was replaced with python3-cryptography ([rhbz#1927404])

[ghissue#123]: https://github.com/ClusterLabs/pcs/issues/123
[rhbz#1720221]: https://bugzilla.redhat.com/show_bug.cgi?id=1720221
[rhbz#1882291]: https://bugzilla.redhat.com/show_bug.cgi?id=1882291
[rhbz#1886342]: https://bugzilla.redhat.com/show_bug.cgi?id=1886342
[rhbz#1896458]: https://bugzilla.redhat.com/show_bug.cgi?id=1896458
[rhbz#1927404]: https://bugzilla.redhat.com/show_bug.cgi?id=1927404

## [0.10.8] - 2021-02-01

### Added
- Support for changing corosync configuration in an existing cluster
  ([rhbz#1457314], [rhbz#1667061], [rhbz#1856397], [rhbz#1774143])
- Command to show structured corosync configuration (see `pcs cluster
  config show` command) ([rhbz#1667066])

### Fixed
- Improved error message with a hint in `pcs cluster cib-push` ([ghissue#241])
- Option --wait was not working with pacemaker 2.0.5+ ([ghissue#260])
- Explicitly close libcurl connections to prevent stalled TCP connections in
  CLOSE-WAIT state ([ghissue#261], [rhbz#1885841])
- Fixed parsing negative float numbers on command line ([rhbz#1869399])
- Removed unwanted logging to system log (/var/log/messages) ([rhbz#1917286])
- Fixed rare race condition in `pcs cluster start --wait` ([rhbz#1794062])
- Better error message when unable to connect to pcsd ([rhbz#1619818])

### Deprecated
- Commands `pcs config import-cman` and `pcs config export
  pcs-commands|pcs-commands-verbose` have been deprecated ([rhbz#1851335])
- Entering values starting with '-' (negative numbers) without '--' on command
  line is now deprecated ([rhbz#1869399])

[ghissue#241]: https://github.com/ClusterLabs/pcs/issues/241
[ghissue#260]: https://github.com/ClusterLabs/pcs/issues/260
[ghissue#261]: https://github.com/ClusterLabs/pcs/issues/261
[rhbz#1457314]: https://bugzilla.redhat.com/show_bug.cgi?id=1457314
[rhbz#1619818]: https://bugzilla.redhat.com/show_bug.cgi?id=1619818
[rhbz#1667061]: https://bugzilla.redhat.com/show_bug.cgi?id=1667061
[rhbz#1667066]: https://bugzilla.redhat.com/show_bug.cgi?id=1667066
[rhbz#1774143]: https://bugzilla.redhat.com/show_bug.cgi?id=1774143
[rhbz#1794062]: https://bugzilla.redhat.com/show_bug.cgi?id=1794062
[rhbz#1851335]: https://bugzilla.redhat.com/show_bug.cgi?id=1851335
[rhbz#1856397]: https://bugzilla.redhat.com/show_bug.cgi?id=1856397
[rhbz#1869399]: https://bugzilla.redhat.com/show_bug.cgi?id=1869399
[rhbz#1885841]: https://bugzilla.redhat.com/show_bug.cgi?id=1885841
[rhbz#1917286]: https://bugzilla.redhat.com/show_bug.cgi?id=1917286


## [0.10.7] - 2020-09-30

### Added
- Support for multiple sets of resource and operation defaults, including
  support for rules ([rhbz#1222691], [rhbz#1817547], [rhbz#1862966],
  [rhbz#1867516], [rhbz#1869399])
- Support for "demote" value of resource operation's "on-fail" option
  ([rhbz#1843079])
- Support for 'number' type in rules ([rhbz#1869399])
- It is possible to set custom (promotable) clone id in `pcs resource create`
  and `pcs resource clone/promotable` commands ([rhbz#1741056])

### Fixed
- Prevent removing non-empty tag by removing tagged resource group or clone
  ([rhbz#1857295])
- Clarify documentation for 'resource move' and 'resource ban' commands with
  regards to the 'lifetime' option.
- Allow moving both promoted and demoted promotable clone resources
  ([rhbz#1875301])

### Deprecated
- `pcs resource [op] defaults <name>=<value>...` commands are deprecated now.
  Use `pcs resource [op] defaults update <name>=<value>...` if you only manage
  one set of defaults, or `pcs resource [op] defaults set` if you manage
  several sets of defaults. ([rhbz#1817547])

[rhbz#1222691]: https://bugzilla.redhat.com/show_bug.cgi?id=1222691
[rhbz#1741056]: https://bugzilla.redhat.com/show_bug.cgi?id=1741056
[rhbz#1817547]: https://bugzilla.redhat.com/show_bug.cgi?id=1817547
[rhbz#1843079]: https://bugzilla.redhat.com/show_bug.cgi?id=1843079
[rhbz#1857295]: https://bugzilla.redhat.com/show_bug.cgi?id=1857295
[rhbz#1862966]: https://bugzilla.redhat.com/show_bug.cgi?id=1862966
[rhbz#1867516]: https://bugzilla.redhat.com/show_bug.cgi?id=1867516
[rhbz#1869399]: https://bugzilla.redhat.com/show_bug.cgi?id=1869399
[rhbz#1875301]: https://bugzilla.redhat.com/show_bug.cgi?id=1875301


## [0.10.6] - 2020-06-11

### Security
- Web UI sends HTTP headers: Content-Security-Policy, X-Frame-Options and
  X-Xss-Protection

### Added
- When creating a cluster, verify the cluster name does not prevent mounting
  GFS2 volumes ([rhbz#1782553])
- An option to run 'pcs cluster setup' in a local mode (do not connect to any
  nodes, save corosync.conf to a specified file) ([rhbz#1839637])
- Support for pacemaker tags. Pcs provides commands for creating and removing
  tags, adding and/or removing IDs to/from tags, and listing current tag
  configuration.  ([rhbz#1684676])
- Support for tag ids in commands resource enable/disable/manage/unmanage
  ([rhbz#1684676])
- `pcs resource [safe-]disable --simulate` has a new option `--brief` to print
  only a list of affected resources ([rhbz#1833114])

### Fixed
- Keep autogenerated IDs of set constraints reasonably short ([rhbz#1387358],
  [rhbz#1824206])
- Pcs is now compatible with Ruby 2.7 and Python 3.8. To achieve this, it newly
  depends on python3-distro package.
- `pcs status` works on remote nodes again (broken since pcs-0.10.4)
  ([rhbz#1830552])
- Fixed inability to create colocation constraint from web ui ([rhbz#1832973])
- Actions going through pcsd no longer time out after 30s (broken since
  pcs-0.10.5) ([rhbz#1833506])


pcs/snmp/pcs_snmp_agent.py
import os
import sys
import logging
import logging.handlers

# pylint: disable=import-error
import pyagentx

import pcs.utils
from pcs.snmp import settings
from pcs.snmp.updaters.v1 import ClusterPcsV1Updater


logger = logging.getLogger("pcs.snmp")
logger.addHandler(logging.NullHandler())


def is_debug():
    debug = os.environ.get("PCS_SNMP_AGENT_DEBUG", "")
    return debug.lower() in ["true", "on", "1"]


def get_update_interval():
    interval = os.environ.get("PCS_SNMP_AGENT_UPDATE_INTERVAL")
    if not interval:
        return settings.DEFAULT_UPDATE_INTERVAL

    def _log_invalid_value(_value):
        logger.warning(
            "Invalid update interval value: '%s' is not >= 1.0", str(_value)
        )
        logger.debug(
            "Using default update interval: %s",
            str(settings.DEFAULT_UPDATE_INTERVAL),
        )

    try:
        interval = float(interval)
    except ValueError:
        _log_invalid_value(interval)
        return settings.DEFAULT_UPDATE_INTERVAL
    if interval <= 1.0:
        _log_invalid_value(interval)
        return settings.DEFAULT_UPDATE_INTERVAL
    return interval


def setup_logging(debug=False):
    level = logging.INFO
    if debug:
        level = logging.DEBUG
        # this is required to enable debug also in the ruby code
        # key '--debug' has to be added
        pcs.utils.pcs_options["--debug"] = debug
    formatter = logging.Formatter(
        "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
    )
    handler = logging.handlers.WatchedFileHandler(
        settings.LOG_FILE, encoding="utf8"
    )
    handler.setLevel(level)
    handler.setFormatter(formatter)
    for logger_name in ["pyagentx", "pcs"]:
        logger_instance = logging.getLogger(logger_name)
        logger_instance.setLevel(level)
        logger_instance.addHandler(handler)


class PcsAgent(pyagentx.Agent):
    def setup(self):
        update_interval = get_update_interval()
        logger.info("Update interval set to: %s", str(update_interval))
        self.register(
            settings.PCS_OID + ".1",
            ClusterPcsV1Updater,
            freq=update_interval,
        )


def main():
    setup_logging(is_debug())
    try:
        agent = PcsAgent()
        agent.start()
    # pylint: disable=broad-except
    except Exception as e:
        print("Unhandled exception: {0}".format(str(e)))
        agent.stop()
        sys.exit(1)
    except KeyboardInterrupt:
        agent.stop()
